ipsec: Fix xfrm output mask

pchaigno opened the pull request:
First commit temporarily uses our netlink fork to be able to set the xfrm output mask. Second commit sets said mask in our xfrm rules. See commit messages for details.

I tested by applying the following diff to our existing test and manually checking the source identity is preserved when making requests between nodes:
```diff
diff --git a/test/k8sT/DatapathConfiguration.go b/test/k8sT/DatapathConfiguration.go
index c11628f67..a91bc17bb 100644
--- a/test/k8sT/DatapathConfiguration.go
+++ b/test/k8sT/DatapathConfiguration.go
@@ -214,7 +214,8 @@ var _ = Describe("K8sDatapathConfig", func() {
 
                        deploymentManager.Deploy(helpers.CiliumNamespace, IPSecSecret)
                        deploymentManager.DeployCilium(map[string]string{
-                               "encryption.enabled": "true",
+                               "encryption.enabled":     "true",
+                               "bpf.monitorAggregation": "none",
                        }, DeployCiliumOptionsAndDNS)
                        validateBPFTunnelMap()
                        Expect(testPodConnectivityAcrossNodes(kubectl)).Should(BeTrue(), "Connectivity test with IPsec between nodes failed")
@@ -736,6 +737,7 @@ func testPodConnectivityAndReturnIP(kubectl *helpers.Kubectl, requireMultiNode b
        if !res.WasSuccessful() {
                return false, targetIP
        }
+       helpers.HoldEnvironment("Let's check the source identity!")
 
        // HTTP connectivity test
        res = kubectl.ExecPodCmd(randomNamespace, srcPod,
```

```release-note
Fix missing packet mark mask that can cause policy deny drops in IPSec configuration. 
```

jrfastab commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -137,7 +140,10 @@ func ipSecReplaceStateOut(remoteIP, localIP net.IP) (uint8, error) {
-	state.OutputMark = linux_defaults.RouteMarkEncrypt
+	state.OutputMark = &netlink.XfrmMark{
+		Value: linux_defaults.RouteMarkEncrypt,
+		Mask:  linux_defaults.IPsecMarkMaskIn,
```
This comment is incorrect see below, leaving for history so I remember what I was thinking.
...
In ipSecReplaceStateOut I believe we want linux_defaults.IPsecMarkMask not the In version of it.

On out we have skb mark like this, 'skb->mark = 0xIDIDFFID' where the 0xID is the 3 bytes of security context split across the mark. And the 0XFF is the encrypt key used to match with IPsec SPI value. So double check please with a test, but I believe its best to drop the "In" above.

jrfastab replied:
Above is wrong (I left though to remind myself I already consider it).

Actually we are just setting the RouteMarkEncrypt=0x0E00 and RouteMarkDecrypt=0x0D00 values. So IPsecMarkMaskIn=0x0F00 is correct, but change to RouteMarkMask so that it reads better and doesn't confuse us later. This is a bit subtle so best to be clear if possible

pchaigno committed:
ipsec: Set mask for xfrm output mark

Commit b6f2e75 set the xfrm output mark for the tunneling mode, but did
not set a mask. As a result, the packet mark is zeroed before writing
the new mark value, clearing our source security context.

In particular, this loss of security context happen when packets are
passed from the lxc devices to cilium_host via the stack. We then
receive a zero source identity in bpf_host and transmit it over the
tunnel to the destination nodes. That causes policy deny drops on remote
nodes when policies are enforced.

The bug can be observed by tracing the packets leaving the containers on
the source node. The source identity becomes unknown when it leaves
cilium_host (-> overlay trace event):

    <- endpoint 684 flow 0x10b975f7 identity 14765->unknown state new ifindex 0 orig-ip 0.0.0.0: 10.0.1.68:41856 -> 10.0.0.206:80 tcp SYN
    -> stack flow 0x10b975f7 identity 14765->30552 state new ifindex 0 orig-ip 0.0.0.0: 10.0.1.68:41856 -> 10.0.0.206:80 tcp SYN
    -> overlay flow 0x10b975f7 identity unknown->unknown state new ifindex cilium_vxlan orig-ip 0.0.0.0: 10.0.1.122 -> 10.0.0.184

This commit fixes the bug by setting the mask appropriately. That
requires us to temporarily use our cilium/netlink fork which has support
for setting the output mask [1], until that support is upstreamed.

1 - https://github.com/cilium/netlink/tree/outputmark-mask
Fixes: b6f2e75 ("cilium: set output mark for tunnel case")
Suggested-by: John Fastabend <john.fastabend@gmail.com>
Signed-off-by: Paul Chaignon <paul@cilium.io>

tklauser reviewed (approved):
Vendoring LGTM, thanks!

I assume we will switch back to use `https://github.com/vishvananda/netlink` once vishvananda/netlink#602 is merged?
